i2c/common/libi2c-msg.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Phoenix-RTOS
+ *
+ * i2c msg client library
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Marek Bialowas
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <sys/msg.h>
+#include <sys/types.h>
+
+#include <stdio.h>
+#include <errno.h>
+
+#include <i2c.h>
+#include <i2c-msg.h>
+
+static struct {
+	oid_t i2c_oid;
+	int initialized;
+} common;
+
+/* opens the device */
+int i2c_init(unsigned int dev_no)
+{
+	char devname[sizeof("/dev/i2cX")];
+	int res = 0;
+
+	sprintf(devname, "/dev/i2c%u", dev_no % 10);
+	res = lookup(devname, NULL, &common.i2c_oid);
+
+	if (res >= 0)
+		common.initialized = 1;
+
+	return res;
+}
+
+
+/* Performs i2c generic write operation to the given slave device. */
+int i2c_busWrite(uint8_t dev_addr, const uint8_t *data, uint32_t len)
+{
+	msg_t msg;
+	int res;
+	i2c_devctl_t *i = (i2c_devctl_t *)msg.i.raw;
+
+	if (!common.initialized)
+		return -EIO;
+
+	msg.type = mtDevCtl;
+
+	msg.i.data = (uint8_t *)data; /* FIXME: dropping const because of broken msg_t declaration */
+	msg.i.size = len;
+	msg.o.data = NULL;
+	msg.o.size = 0;
+
+	i->type = i2c_devctl_bus_write;
+	i->dev_addr = dev_addr;
+
+	if ((res = msgSend(common.i2c_oid.port, &msg)) < 0)
+		return res;
+
+	return msg.o.io.err;
```

Comment:
If you use devctl, don't use fields for IO operations. For `mtDevCtl` there are only `raw` fields available. You should pass error code in a custom struct (so maybe just add `err` field to `i2c_devctl_t` and pass it in both `msg.i.raw `and `msg.o.raw`?)

Reply:
no docs and does it really matter? As we're using `msg` fields directly everywhere, I don't consider their placement an implementation detail.

Will change to some other solution but I believe we were talking about providing one common `err` field in `msg.o` also instead of repeating it in every union member...

Reply:
introduced i/o union of structs in `i2c_devctl_t`